interpretation/train_classifier.py

```diff
@@ -0,0 +1,219 @@
+        cvresult = xgb.cv(xgb_param, xgtrain, num_boost_round=model.get_params()['n_estimators'],
+                          folds=folds, nfold=cv_folds, verbose_eval=True,
+                          metrics='auc', early_stopping_rounds=early_stopping_rounds, stratified=True)
+        model.set_params(n_estimators=cvresult.shape[0])
+        test_score = cvresult['test-auc-mean'].iloc[-1]
+        print('# of estimators: %d' % cvresult.shape[0])
+
+    # retrain the model with the whole dataset
+    model.fit(dtrain[features], dtrain[target], verbose=True, eval_metric='auc')
+    train_yhat = model.predict(dtrain[features])
+    train_score = metrics.accuracy_score(dtrain[target].tolist(), train_yhat)
+    if cv:
+        score = test_score
+        print("AUC Score (Test): %f" % score)
+    else:
+        score = train_score
+        print("Accuracy : %.4g" % score)
+    return model, score
+
+def tune_cls_para(model, train, features, target, param_test, cv=None, n_jobs=8):
+    """Tune classifier parameters with sklearn GridSearchCV
+    :param object model: estimator object
+    :param dataframe train: training data with rows being samples and columns being features
+    :param list features: column names of features
+    :param str target: column name of the target
+    :param dict or list of dictionaries param_test: Dictionary with parameters names (string)
+        as keys and lists of
+        parameter settings to try as values, or a list of such
+        dictionaries, in which case the grids spanned by each dictionary
+        in the list are explored. This enables searching over any sequence
+        of parameter settings.
+    :param int, cross-validation generator or an iterable cv: cross-validation splitting strategy.
+    :param int or None n_jobs: Number of jobs to run in parallel.
+    :return object model: estimator object with optimal parameters.
+    """
+    gsearch = GridSearchCV(estimator=model, param_grid=param_test,
+                           scoring='roc_auc', n_jobs=n_jobs, iid=False, verbose=3,
+                           cv=cv)
+    gsearch.fit(train[features], train[target])
+    model.set_params(**gsearch.best_params_)
+    means = gsearch.cv_results_['mean_test_score']
+    print(means)
+    stds = gsearch.cv_results_['std_test_score']
+    for mean, std, params in zip(means, stds, gsearch.cv_results_['params']):
+        print("%0.4f (+/-%0.04f) for %r"
+              % (mean, std * 2, params))
+    print('best CV parameters:')
+    print(gsearch.best_params_)
+    print('best CV score: %f' % gsearch.best_score_)
+    return model
+
+def plot_xgb_fscore(model, output_dir, output_fname):
+    """plot xgboost feature importance"""
+    fig = plt.figure()
+    fig.set_size_inches((6, 6))
+    xgb.plot_importance(model, importance_type='gain')
+    plt.savefig(os.path.join(output_dir, ''.join([output_fname, '.jpg'])),
+                dpi=300, bbox_inches='tight')
+    plt.close()
+
+if __name__ == '__main__':
```

Comment:
You should put most of your code in functions or classes, not in __main__. See e.g. 
https://realpython.com/python-main-function/

Reply:
I moved most of the code to the function `main()` now. This is intended to be more a script demonstrating how to build a classifier for pysero output rather a general pipeline for training classifier (or at least not at the moment). Perhaps we can refactor this later?